determinism.md

```diff
@@ -0,0 +1,171 @@
+GetDeterministicKey
+===
+
+**This is meant to be the text of the issue I will eventually file for the API**
+
+## Background and Motivation
+The `Compilation` type is fully deterministic meaning that given the same inputs
+(`SyntaxTree`, `CompilationOptions`, etc ...) it will produce the same output. By 
+that it means it will produce byte for byte equivalent binaries and the same 
+diagnostics. This is extremely valuable in infrastructure because it allows 
+for caching and opens the door for efficient distributed processing.
+
+At the moment this is hard to leverage because consumers can't determine if 
+two `Compilation` instances are equivalent for the purposes of determinism. 
+Customers have to resort to hand written comparisons which requires a fairly 
+intimate knowledge of the compiler (for example knowing what does and does not
+impact determinism). Such solutions are not version tolerant; every time the 
+compiler adds a new property that impacts determinism the solution must be 
+updated. Even when proper equality checks are in place this does not help 
+distributed computing where equality must be decided across different processes.
+
+The motivation here is to provide an API that returns a string based key for a
+given `Compilation` such that for two equivalent `Compilation` instances the
+key will also be equivalent. This will allow customers who want to leverage 
+deterministic caching and communication to have a simple, and transmittable,
+way to full describe the `Compilation` they are processing.
+
+## Proposed API
+
+```diff
+namespace Microsoft.CodeAnalysis
+{
+    public class Compilation
+    {
++       public string GetDeterministicKey(
++           ImmutableArray<AdditionalText> additionalTexts = default,
++           ImmutableArray<DiagnosticAnalyzer> analyzers = default,
++           ImmutableArray<ISourceGenerator> generators = default,
++           EmitOptions? emitOptions = null,
++           DeterministicKeyOptions options = DeterministicKeyOptions.Default)
+    }
+
+    internal enum DeterministicKeyOptions
+    {
++       /// <summary>
++       /// The default is to include all inputs to the compilation which impact the output of the 
++       /// compilation: binaries or diagnostics.
++       /// </summary>
++       Default = 0b0,
+
++       /// <summary>
++       /// Ignore all file paths, but still include file names, in the deterministic key.
++       /// </summary>
++       IgnorePaths = 0b0001,
```

Comment:
File names are a gray area of determinism. In the strict sense yes they matter because it impacts data written to the PDB, `[CallerFileInfo]` and certain diagnostic messages. A significant amount of customers are likely not concerned with these differences though. 

Consider back when we had caching in our CI system. The specific path that our code was checked out to was not material to our build and test caching. Or stated differently the same commit run through CI but checked out into two different directories always had the same outcome in terms of build failures and tests. Because source paths matter though we had to jump through hoops in our build and caching logic to normalize out the path differences. This API provides a very simple way to say 

> Give me the key that is independent of the location where the code is built from

The reason I kept the names vs. just deleting the paths entirely is that there is one other case that matters for determinism here. Bit of an edge case but consider that the compiler will error on the command line if you pass the same file twice. Keeping the file name means that we catch that case in the key. 

File names have some other tricky points though that we need to consider and account for. For example "aux.txt" is a valid file name on Windows 11 but not Windows 10. So to have a fully correct key we need to encode the OS version into the key.